tests/qtrajectory_test.cc

```diff
@@ -0,0 +1,439 @@
+
+  auto normal = KrausOperator<types::Gate>::kNormal;
+
+  using M = Cirq::MatrixGate1<fp_type>;
+
+  ncircuit.push_back(
+      {{normal, 0, p1, {M::Create(time, 0, {t1, 0, 0, 0, 0, 0, r1, 0})}},
+       {normal, 0, p2, {M::Create(time, 0, {r2, 0, 0, 0, 0, 0, t2, 0})}},
+       {normal, 0, p3, {M::Create(time, 0, {0, 0, s1, 0, 0, 0, 0, 0})}},
+       {normal, 0, p3, {M::Create(time, 0, {0, 0, 0, 0, s2, 0, 0, 0})}}});
+  ncircuit.push_back(
+      {{normal, 0, p1, {M::Create(time, 1, {t1, 0, 0, 0, 0, 0, r1, 0})}},
+       {normal, 0, p2, {M::Create(time, 1, {r2, 0, 0, 0, 0, 0, t2, 0})}},
+       {normal, 0, p3, {M::Create(time, 1, {0, 0, s1, 0, 0, 0, 0, 0})}},
+       {normal, 0, p3, {M::Create(time, 1, {0, 0, 0, 0, s2, 0, 0, 0})}}});
+}
+
+template <typename AddNoise>
+types::NoisyCircuit GenerateNoisyCirquit(double p, AddNoise&& add_noise) {
+  using fp_type = types::Gate::fp_type;
+
+  types::NoisyCircuit ncircuit;
+  ncircuit.reserve(24);
+
+  using Hd = Cirq::H<fp_type>;
+  using IS = Cirq::ISWAP<fp_type>;
+  using Rx = Cirq::rx<fp_type>;
+  using Ry = Cirq::ry<fp_type>;
+
+  auto normal = KrausOperator<types::Gate>::kNormal;
+
+  ncircuit.push_back({{normal, 1, 1.0, {Hd::Create(0, 0)}}});
+  ncircuit.push_back({{normal, 1, 1.0, {Hd::Create(0, 1)}}});
+  add_noise(1, p, ncircuit);
+  ncircuit.push_back({{normal, 1, 1.0, {IS::Create(2, 0, 1)}}});
+  add_noise(3, p, ncircuit);
+  ncircuit.push_back({{normal, 1, 1.0, {Rx::Create(4, 0, 0.7)}}});
+  ncircuit.push_back({{normal, 1, 1.0, {Ry::Create(4, 1, 0.1)}}});
+  add_noise(5, p, ncircuit);
+  ncircuit.push_back({{normal, 1, 1.0, {IS::Create(6, 0, 1)}}});
+  add_noise(7, p, ncircuit);
+  ncircuit.push_back({{normal, 1, 1.0, {Ry::Create(8, 0, 0.4)}}});
+  ncircuit.push_back({{normal, 1, 1.0, {Rx::Create(8, 1, 0.7)}}});
+  add_noise(9, p, ncircuit);
+  ncircuit.push_back({{normal, 1, 1.0, {IS::Create(10, 0, 1)}}});
+  add_noise(11, p, ncircuit);
+  ncircuit.push_back({{KrausOperator<types::Gate>::kMeasurement, 1, 1.0,
+                       {gate::Measurement<types::Gate>::Create(12, {0, 1})}}});
+  add_noise(13, p, ncircuit);
+
+  return ncircuit;
+}
+
+void Run1(const types::NoisyCircuit& ncircuit,
+          const std::vector<double>& expected_results) {
+  unsigned num_qubits = 2;
+  unsigned num_reps = 20000;
+
+  auto measure = [](uint64_t r, const types::State& state,
+                    const std::vector<uint64_t>& stat,
+                    std::vector<unsigned>& histogram) {
+    ASSERT_EQ(stat.size(), 1);
+    ++histogram[stat[0]];
+  };
+
+  std::vector<unsigned> histogram(1 << num_qubits, 0);
+
+  types::QTSimulaltor::Parameter param;
+  param.collect_mea_stat = true;
+
+  EXPECT_TRUE(types::QTSimulaltor::Run(param, num_qubits, ncircuit,
+                                       0, num_reps, measure, histogram));
+
+  for (std::size_t i = 0; i < histogram.size(); ++i) {
+    EXPECT_NEAR(double(histogram[i]) / num_reps, expected_results[i], 0.005);
```

Comment:
Requiring accuracy within 0.5% sounds good to me - do you have a sense for how close we are to this boundary? Using fixed seeds should prevent flakes, but it'd be nice to know whether future changes to the operator-selection code are at risk of triggering a failure here.

Reply:
We are close to this boundary. The largest difference is about 0.00465 for 20000 repetitions. The accuracy depends on the number of repetitions and it is of course much better for, say, 10000000 repetitions. However, this leads to much longer test run times. Yes, future changes to the operator-selection code are at some risk of triggering a failure here. This can also be fixed by increasing the number of repetitions.